service/src/main/java/bio/terra/workspace/service/workspace/flight/DeleteGcpContextFlight.java

```diff
@@ -5,24 +5,42 @@
 import bio.terra.stairway.RetryRule;
 import bio.terra.stairway.RetryRuleExponentialBackoff;
 import bio.terra.workspace.common.utils.FlightBeanBag;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.job.JobMapKeys;
+import bio.terra.workspace.service.workspace.model.CloudPlatform;
+import java.util.UUID;
 
 /** A {@link Flight} for deleting a Google cloud context for a workspace. */
 public class DeleteGcpContextFlight extends Flight {
   private static final int INITIAL_INTERVALS_SECONDS = 1;
   private static final int MAX_INTERVAL_SECONDS = 8;
   private static final int MAX_OPERATION_TIME_SECONDS = 5 * 60;
+  private static final CloudPlatform CLOUD_PLATFORM = CloudPlatform.GCP;
 
   public DeleteGcpContextFlight(FlightMap inputParameters, Object applicationContext) {
     super(inputParameters, applicationContext);
 
     FlightBeanBag appContext = FlightBeanBag.getFromObject(applicationContext);
+    UUID workspaceId =
+        UUID.fromString(inputParameters.get(WorkspaceFlightMapKeys.WORKSPACE_ID, String.class));
+    AuthenticatedUserRequest userRequest =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
 
     RetryRule retryRule =
         new RetryRuleExponentialBackoff(
             INITIAL_INTERVALS_SECONDS, MAX_INTERVAL_SECONDS, MAX_OPERATION_TIME_SECONDS);
 
+    addStep(
+        new ListControlledResourceIdsStep(appContext.getResourceDao(), workspaceId, CLOUD_PLATFORM),
+        retryRule);
+    addStep(
```

Comment:
There is a potential Sam leak here if a new controlled resource is created after the enumeration.
The Sam step will miss it, but the metadata step will delete it. For now, would you make a comment to that effect. I am thinking that we could use locking to close the window. There is already a ticket PF-555 you can reference.

Is there a reason to split the List and the DeleteControlledSamResourceStep into two steps?
It would be idempotent to read the list in memory and apply it to Sam in one go. If the step failed and restarted, it would read the list again and apply it to Sam again. We would have to tolerate "not found" in the Sam call.

Reply:
Added TODO referencing PF-555

No strong reason to separate the two other than I thought of them as distinct operations. Given that this saves serializing/deserializing a list of complex objects, that seems like a better approach. 
`SamService.deleteControlledResource` internally handles not found errors, so we don't need to manage those in steps.